### DISCUSSION TITLE
How to disable `test.require` stage

### DISCUSSION THREAD

**Discussion by LecrisUT** [teemtee/tmt#2569] [Q&A]
If necessary through `tmt run` that would work as well

**Comment by lukaszachy** [comment]
You mean stop tmt run to install require/recommend during 'prepare' step? Not possible ATM :(

But shouldn't be that hard to implement

**Reply by LecrisUT** [reply]
Yeah, but it would likely clash with #2557, so we should wait for that first?

**Reply by lukaszachy** [reply]
Waiting doesn't hurt (too much on the plate already) but it should conflict too much. Whole idea about such flag would be to stop 'install' to be injected automatically (and usual troubles about name of such option and where it should live...)

**Reply by happz** [reply]
Maybe an option in `prepare` subcommand? There's already `-x/--exit-first` in `execute`...

The smallest incision would be skipping these two loops, https://github.com/teemtee/tmt/blob/main/tmt/steps/prepare/__init__.py#L313 and https://github.com/teemtee/tmt/blob/main/tmt/steps/prepare/__init__.py#L329. A bigger patch might make the whole collection of requirements/recommends optional, but it still might be something worth preserving, e.g. for logging purposes - "this is what we would install, but you told us not to do so".

**Reply by lukaszachy** [reply]
> Maybe an option in prepare subcommand? There's already -x/--exit-first in execute...

Very likely, we should make it available just for prepare itself (not inherited to the plugins) as it doesn't make sense of shell/ansible plugins. How about `-s/--skip-install` ?

> "this is what we would install, but you told us not to do so".

+1

**Reply by psss** [reply]
Common `prepare` step option sounds ok. On the other hand, something like `--skip-require` or `--skip-recommend` sounds irrelevant for most `prepare` plugins. Perhaps, there could be an explicit way how to adjust only the require packages installation? Something like `prepare --how require --skip` or `prepare --how special --skip-require --skip-recommend`? Not sure.

**Reply by LecrisUT** [reply]
How does this work when there are already prepare steps in the plan that we don't want to override?

**Reply by lukaszachy** [reply]
There is `--update` which 'updates' just selected phase name

**Reply by LecrisUT** [reply]
I was thinking about that, but what would we pass to it to select the auto-generated prepare step inserted by `test.require`?

**Reply by happz** [reply]
That's why I would prefer `prepare`-level options, `prepare --skip-require --skip-recommend`, or `prepare --how special --skip-require --skip-recommend`, i.e. not depending on the user being aware of a special phase name.

**Reply by LecrisUT** [reply]
How would that work in `.fmf`, e.g. if we want to automatically include it via `adjust` and `context.distro`

**Reply by happz** [reply]
Hmm, good point... I'm afraid `adjust` won't work here, because `adjust` applies when tmt loads fmf files, then `adjust` is evaluated, then tmt gets the resulting tree & converts it into internal objects like tests and stories, and way later, tmt adds a specially constructed `install` phase into `prepare` step to install stuff. This happens so late after `adjust` that it has no chance of affecting this. When `adjust` happens, this special phase does not exist, because it's not written in fmf files.

**Reply by LecrisUT** [reply]
I was thinking that there is an even more fundamental issue that `prepare` can be an array, at which point we cannot add a top-level key-value and keep it yaml parseable. But I guess `adjust` would be able to overcome this?

As for the order, couldn't the `skip_require` field be checked right before injecting the `install.require`?

**Reply by happz** [reply]
> I was thinking that there is an even more fundamental issue that prepare can be an array, at which point we cannot add a top-level key-value and keep it yaml parseable.

Correct, but that could be solved by adding "special" plugin representing this operation, which would own these options and which one would be able to disable. It'd be equivalent to `prepare --how special --skip-require --skip-recommend`.

`adjust` should work in this case as well, are we heading towards something like this? @psss @lukaszachy WDYT?

```yaml
prepare:
  - how: install
    packages:
      - foo

  - how: ansible
    playbook /foo.yaml

  # the special one, represents the phase added by `prepare`
  - how: requires
    skip-require: true
    skip-recommend: false

    adjust:
      when: distro == rhel-9
      enabled: true
```

**Reply by lukaszachy** [reply]
At the moment `adjust` is a plan level key, so above example can't work. What could work (with current adjust) is 
```
adjust:
  when: distro != rhel-9 
  prepare+:
     - how: requires
       skip-require: true
```

But having adjust closer to the step phase would be really nice to have.

**Reply by lukaszachy** [reply]
Another thought though: Karel had rfe to install require before each test so isn't it more 'run' property and the fact it happens in prepare plugin is just an implementation detail? So something like `tmt run --skip-install-require` or `tmt run --install [prepare|execute|never]` ?

We can create place to store such option in the plan.fmf somehow (and be subject of adjust)
